File: 04-batch-performance/ai/java/src/main/java/com/treinamento/batch/Main.java
package com.treinamento.batch;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

/**
 * CLI entry point for the batch transaction processor.
 *
 * <h2>Usage</h2>
 * <pre>
 *   mvn exec:java -Dexec.args="data/transactions_10k.csv"
 *   mvn exec:java -Dexec.args="data/transactions_10k.csv --generate 10000"
 * </pre>
 */
public class Main {

    public static void main(String[] args) {
        if (args.length == 0) {
            System.err.println("Usage: batch-processor <transactions.csv> [--generate <count>]");
            System.err.println();
            System.err.println("Examples:");
            System.err.println("  mvn exec:java -Dexec.args=\"data/transactions_10k.csv\"");
            System.err.println("  mvn exec:java -Dexec.args=\"data/transactions_10k.csv --generate 10000\"");
            System.exit(1);
        }

        Path csvPath = Path.of(args[0]);

        // Optional: auto-generate the file if it is missing and --generate is supplied
        if (!Files.exists(csvPath)) {
            if (args.length >= 3 && "--generate".equals(args[1])) {
                int count = Integer.parseInt(args[2]);
                autoGenerate(csvPath, count);
            } else {
                System.err.println("File not found: " + csvPath);
                System.err.println("Hint: add --generate <count> to auto-generate it, e.g.:");
                System.err.println("  mvn exec:java -Dexec.args=\"" + args[0]
                        + " --generate 10000\"");
                System.exit(1);
            }
        }

        // Read transactions
        System.out.println("Reading transactions from: " + csvPath.toAbsolutePath());
        List<Transaction> transactions;
        try {
            List<String> lines = Files.readAllLines(csvPath);
            transactions = new java.util.ArrayList<>();
            for (String line : lines) {
                String trimmed = line.trim();
                if (trimmed.isEmpty() || trimmed.startsWith("id,")) continue;
                try {
                    transactions.add(Transaction.fromCsv(trimmed));
                } catch (Exception e) {
                    // skip malformed rows
                }
            }
        } catch (IOException e) {
            System.err.println("I/O error reading " + csvPath + ": " + e.getMessage());
            System.exit(2);
            return;
        }
        System.out.printf("Loaded %,d transactions.%n%n", transactions.size());

        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        System.out.println("Inicio: " + LocalDateTime.now().format(dtf));
        System.out.println();

        BatchProcessor processor = new BatchProcessor();
        long startTime = System.currentTimeMillis();
        ProcessingResult result = processor.processTransactions(transactions);
        double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

        System.out.println();
        System.out.println("Termino: " + LocalDateTime.now().format(dtf));
        System.out.println();

        // -------------------------------------------------------------------------
        // Summary output — matches Bun format exactly
        // -------------------------------------------------------------------------
        System.out.println("=".repeat(60));
        System.out.println("  BATCH PROCESSING RESULTS");
        System.out.println("=".repeat(60));
        System.out.printf(java.util.Locale.US, "  Processed   : %,d%n", result.processed());
        System.out.printf(java.util.Locale.US, "  Duplicates  : %,d%n", result.duplicates());
        System.out.printf(java.util.Locale.US, "  Rejected    : %,d%n", result.rejected());
        System.out.printf(java.util.Locale.US, "  Fraud rings : %,d%n", result.fraudRings());
        System.out.printf(java.util.Locale.US, "  Time        : %.3fs%n", elapsed);
        System.out.println("=".repeat(60));

        // Category Report
        System.out.println();
        System.out.println("Category Report:");
        System.out.printf("  %-16s  %8s  %14s  %12s%n",
            "Category", "Count", "Total", "Average");
        System.out.println("  " + "-".repeat(54));
        for (ProcessingResult.CategoryReport stat : result.report()) {
            System.out.printf(java.util.Locale.US, "  %-16s  %8d  %14.2f  %12.2f%n",
                stat.category(), stat.count(), stat.total(), stat.average());
        }

        System.out.println();
    }

    // -------------------------------------------------------------------------
    // Helper
    // -------------------------------------------------------------------------

    private static void autoGenerate(Path csvPath, int count) {
        try {
            if (csvPath.getParent() != null) {
                Files.createDirectories(csvPath.getParent());
            }
            System.out.printf("Generating %,d transactions → %s%n", count, csvPath);
            DataGenerator.generate(count, csvPath);
            System.out.println("Generation complete.");
            System.out.println();
        } catch (IOException e) {
            System.err.println("Failed to generate dataset: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: 04-batch-performance/ai/kotlin/src/main/kotlin/com/treinamento/batch/Main.kt
package com.treinamento.batch

import java.io.File
import java.text.NumberFormat
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * CLI entry point for the batch processor.
 *
 * Usage:
 *   ./gradlew run --args="<csv-file>"
 *   ./gradlew run --args="--generate <path> <count>"
 *   ./gradlew run --args="--generate-and-run <path> <count>"
 */
fun main(args: Array<String>) {
    when {
        args.isEmpty() -> {
            printUsage()
            return
        }

        args[0] == "--generate" -> {
            if (args.size < 3) { printUsage(); return }
            val path  = args[1]
            val count = args[2].toIntOrNull() ?: run {
                System.err.println("ERROR: <count> must be an integer, got '${args[2]}'")
                return
            }
            DataGenerator.generateToPath(count, path)
        }

        args[0] == "--generate-and-run" -> {
            if (args.size < 3) { printUsage(); return }
            val path  = args[1]
            val count = args[2].toIntOrNull() ?: run {
                System.err.println("ERROR: <count> must be an integer, got '${args[2]}'")
                return
            }
            DataGenerator.generateToPath(count, path)
            processFile(path)
        }

        else -> processFile(args[0])
    }
}

// -- File processing ----------------------------------------------------------

private fun processFile(path: String) {
    val file = File(path)
    if (!file.exists()) {
        System.err.println("Error reading file \"$path\": File not found")
        System.err.println("Hint: generate a dataset first with --generate <path> <count>")
        return
    }

    println("Reading transactions from: $path")

    val transactions: List<Transaction> = try {
        file.bufferedReader().useLines { lines ->
            lines
                .map { it.trim() }
                .filter { it.isNotEmpty() && !it.startsWith("id,") }
                .map { line ->
                    runCatching { Transaction.fromCsvLine(line) }.getOrNull()
                }
                .filterNotNull()
                .toList()
        }
    } catch (ex: Exception) {
        System.err.println("Error reading file \"$path\": ${ex.message}")
        return
    }

    val nf = NumberFormat.getNumberInstance(Locale.US)
    println("Loaded ${nf.format(transactions.size)} transactions.\n")

    if (transactions.isEmpty()) {
        println("No valid transactions found - nothing to process.")
        return
    }

    val dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")

    println("Inicio: ${LocalDateTime.now().format(dtf)}\n")

    val processor = BatchProcessor()

    val startTime = System.nanoTime()
    val result = processor.process(transactions)
    val elapsed = (System.nanoTime() - startTime) / 1_000_000_000.0
    val elapsedStr = "%.3f".format(java.util.Locale.US, elapsed)

    println("\nTermino: ${LocalDateTime.now().format(dtf)}\n")

    // Summary output
    println("=".repeat(60))
    println("  BATCH PROCESSING RESULTS")
    println("=".repeat(60))
    println("  Processed   : ${nf.format(result.processed)}")
    println("  Duplicates  : ${nf.format(result.duplicates)}")
    println("  Rejected    : ${nf.format(result.rejected)}")
    println("  Fraud rings : ${nf.format(result.fraudRings)}")
    println("  Time        : ${elapsedStr}s")
    println("=".repeat(60))

    println("\nCategory Report:")
    println("  " + listOf(
        "Category".padEnd(16),
        "Count".padStart(8),
        "Total".padStart(14),
        "Average".padStart(12)
    ).joinToString("  "))
    println("  " + "-".repeat(54))
    for (stat in result.report) {
        println("  " + listOf(
            stat.category.padEnd(16),
            stat.count.toString().padStart(8),
            "%.2f".format(java.util.Locale.US, stat.total).padStart(14),
            "%.2f".format(java.util.Locale.US, stat.average).padStart(12)
        ).joinToString("  "))
    }

    println()
}

// -- Help ---------------------------------------------------------------------

private fun printUsage() {
    println("""
        Batch Transaction Processor - Performance Challenge

        USAGE:
          ./gradlew run --args="<csv-file>"
              Process an existing CSV file.

          ./gradlew run --args="--generate <path> <count>"
              Generate a synthetic CSV with <count> transactions.

          ./gradlew run --args="--generate-and-run <path> <count>"
              Generate and immediately process.

        EXAMPLES:
          ./gradlew run --args="--generate data/transactions_10k.csv 10000"
          ./gradlew run --args="data/transactions_10k.csv"
          ./gradlew run --args="--generate-and-run data/transactions_1k.csv 1000"
    """.trimIndent())
}
